### PR TITLE
Add model-hub tests to check wrap_model 

### PR DESCRIPTION
**Pull request by AlexanderDokuchaev**
### Changes

- Add models_hub_tests to check `wrap_model`,  like https://github.com/openvinotoolkit/openvino/tree/master/tests/model_hub_tests/torch_tests
- Add `Model Hub` github action to manual run 

### Related tickets

124852

### Tests

tests/torch/models_hub_test

### CODE REVIEW

**Comment by AlexanderDokuchaev** [comment]
Github action can be run before merge to develop. 
Checked on default host (with low limit of RAM), test runs but killed by OOM.

**Review by vshampor — CHANGES_REQUESTED** [review]

**Comment by AlexanderDokuchaev** [comment]
https://github.com/openvinotoolkit/nncf/actions/runs/6930010257/job/18848850899?pr=2275

File: tests/torch/models_hub_test/common.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+            elif len(splitted) == 2:
+                model_name, model_link = splitted
+            elif len(splitted) == 4:
+                model_name, model_link, mark, reason = splitted
+                if model_link == "none":
+                    model_link = None
+                assert mark in ["skip", "xfail"], "Incorrect failure mark for model info {}".format(model_info)
+            else:
+                raise RuntimeError(f"Incorrect model info `{model_info}`. It must contain either 1, 2 or 3 fields.")
+            models.append(ModelInfo(model_name, model_link, mark, reason))
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply tests/torch/models_hub_test/common.py]: The comment has been reviewed and no further action will be taken. 
@MaximProshin

File: tests/torch/models_hub_test/test_hf_transformers.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+        auto_processor = None
+        model = None
+        example = None
+        try:
+            auto_model = mi.transformersInfo["auto_model"]
+            if "processor" in mi.transformersInfo:
+                auto_processor = mi.transformersInfo["processor"]
+        except:  # noqa: E722
+            auto_model = None
+        if "clip_vision_model" in mi.tags:
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply tests/torch/models_hub_test/test_hf_transformers.py:76]: The comment has been reviewed and no further action will be taken. 
@MaximProshin

File: tests/torch/models_hub_test/test_torchvision_models.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    Download video and return frames.
+    Using free video from pexels.com, credits go to Pavel Danilyuk.
+    Initially used in https://pytorch.org/vision/stable/auto_examples/plot_optical_flow.html
+    """
+    from pathlib import Path
+    from urllib.request import urlretrieve
+
+    from torchvision.io import read_video
+
+    video_url = "https://download.pytorch.org/tutorial/pexelscom_pavel_danilyuk_basketball_hd.mp4"
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply tests/torch/models_hub_test/test_torchvision_models.py:41]: The comment has been reviewed and no further action will be taken. 
@MaximProshin

File: tests/torch/models_hub_test/__init__.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2023 Intel Corporation
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply tests/torch/models_hub_test/__init__.py:1]: The comment has been reviewed and no further action will be taken. 
@MaximProshin

File: tests/torch/models_hub_test/common.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+def get_model_params(file_name: Path) -> List[Union[ModelInfo, ParameterSet]]:
+    model_list = get_models_list(file_name)
+    params = []
+    for mi in model_list:
+        if mi.mark == "skip":
+            params.append(pytest.param(mi, marks=pytest.mark.skip(reason=mi.reason)))
+        elif mi.mark == "xfail":
+            params.append(pytest.param(mi, marks=pytest.mark.xfail(reason=mi.reason)))
+        else:
+            params.append(mi)
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply tests/torch/models_hub_test/common.py:141]: The comment has been reviewed and no further action will be taken. 
@MaximProshin

File: tests/torch/models_hub_test/common.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+import pytest
+import torch
+from _pytest.mark import ParameterSet
+
+from nncf.common.graph import NNCFGraph
+from nncf.torch.model_creation import wrap_model
+
+
+@pytest.mark.models_hub
+class BaseTestModel(ABC):
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply tests/torch/models_hub_test/common.py:31]: The comment has been reviewed and no further action will be taken. 
@MaximProshin

File: tests/torch/models_hub_test/test_hf_transformers.py [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+@pytest.fixture(scope="class", autouse=True)
+def fixture_image(request):
+    import requests
+    from PIL import Image
+
+    url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+    request.cls.image = Image.open(requests.get(url, stream=True).raw)
+
+
+class TestTransformersModel(BaseTestModel):
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply tests/torch/models_hub_test/test_hf_transformers.py:62]: The comment has been reviewed and no further action will be taken. 
@MaximProshin

File: tests/torch/models_hub_test/test_hf_transformers.py [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+            for k, v in example:
+                if k in names_set:
+                    new_example[k] = v
+        return new_example
+    except:  # noqa: E722
+        return example
+
+
+@pytest.fixture(scope="class", autouse=True)
+def fixture_image(request):
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply tests/torch/models_hub_test/test_hf_transformers.py:54]: The comment has been reviewed and no further action will be taken. 
@MaximProshin

File: tests/torch/models_hub_test/common.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        example_input = None
+        if isinstance(example, (list, tuple)):
+            example_input = tuple([torch.tensor(x) for x in example])
+        elif isinstance(example, dict):
+            example_input = {k: torch.tensor(v) for k, v in example.items()}
+        assert example_input is not None
+
+        nncf_model = wrap_model(fw_model, example_input)
+
+        self.check_graph(nncf_model.nncf.get_original_graph())
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply tests/torch/models_hub_test/common.py:74]: The comment has been reviewed and no further action will be taken. 
@MaximProshin

File: tests/torch/models_hub_test/test_speech_transformer.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+import torch
+
+from tests.torch.models_hub_test.common import BaseTestModel
+
+
+class TestSpeechTransformerModel(BaseTestModel):
+    def load_model(self, model_name: str):
+        self.repo_dir = tempfile.TemporaryDirectory()
+        os.system(f"git clone https://github.com/mvafin/Speech-Transformer.git {self.repo_dir.name}")
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply tests/torch/models_hub_test/test_speech_transformer.py]: The comment has been reviewed and no further action will be taken. 
@MaximProshin

File: tests/torch/models_hub_test/test_speech_transformer.py [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+        subprocess.check_call(["git", "checkout", "071eebb7549b66bae2cb93e3391fe99749389456"], cwd=self.repo_dir.name)
+
+        sys.path.append(self.repo_dir.name)
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply tests/torch/models_hub_test/test_speech_transformer.py]: The comment has been reviewed and no further action will be taken. 
@MaximProshin

File: tests/torch/models_hub_test/test_timm.py [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+        "large", "xlarge", "xxlarge", "huge", "gigantic", "giant", "enormous", "xs", "xxs", "s", "m", "l", "xl"
+    }  # fmt: skip
+    for name in timm_list:
+        # first: remove datasets
+        name_parts = name.split(".")
+        _name = "_".join(name.split(".")[:-1]) if len(name_parts) > 1 else name
+        # second: remove sizes
+        name_set = set([n for n in _name.split("_") if not n.isnumeric()])
+        name_set = name_set.difference(ignore_set)
+        name_join = "_".join(name_set)
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply tests/torch/models_hub_test/test_timm.py:37]: The comment has been reviewed and no further action will be taken. 
@MaximProshin

File: tests/torch/models_hub_test/test_timm.py [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+    return filtered_list
+
+
+def get_all_models() -> list:
+    m_list = timm.list_pretrained()
+    return filter_timm(m_list)
+
+
+class TestTimmModel(BaseTestModel):
+    def load_model(self, model_name: str):
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply tests/torch/models_hub_test/test_timm.py]: The comment has been reviewed and no further action will be taken. 
@MaximProshin

File: tests/torch/models_hub_test/test_speech_transformer.py [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+import sys
+import tempfile
+
+import torch
+
+from tests.torch.models_hub_test.common import BaseTestModel
+
+
+class TestSpeechTransformerModel(BaseTestModel):
+    def load_model(self, model_name: str):
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply tests/torch/models_hub_test/test_speech_transformer.py]: The comment has been reviewed and no further action will be taken. 
@MaximProshin

File: tests/torch/models_hub_test/test_hf_transformers.py [hunk 15]
```diff
@@ -0,0 +1,302 @@
...
+    import requests
+    from PIL import Image
+
+    url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+    request.cls.image = Image.open(requests.get(url, stream=True).raw)
+
+
+class TestTransformersModel(BaseTestModel):
+    def load_model(self, name: str):
+        torch.manual_seed(0)
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply tests/torch/models_hub_test/test_hf_transformers.py:64]: The comment has been reviewed and no further action will be taken. 
@MaximProshin

File: tests/torch/models_hub_test/test_hf_transformers.py [hunk 16]
```diff
@@ -0,0 +1,302 @@
...
+        return example
+
+
+@pytest.fixture(scope="class", autouse=True)
+def fixture_image(request):
+    import requests
+    from PIL import Image
+
+    url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+    request.cls.image = Image.open(requests.get(url, stream=True).raw)
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply tests/torch/models_hub_test/test_hf_transformers.py:59]: The comment has been reviewed and no further action will be taken. 
@MaximProshin

File: tests/torch/models_hub_test/test_speech_transformer.py [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+        self.repo_dir = tempfile.TemporaryDirectory()
+        os.system(f"git clone https://github.com/mvafin/Speech-Transformer.git {self.repo_dir.name}")
+        subprocess.check_call(["git", "checkout", "071eebb7549b66bae2cb93e3391fe99749389456"], cwd=self.repo_dir.name)
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply tests/torch/models_hub_test/test_speech_transformer.py]: The comment has been reviewed and no further action will be taken. 
@MaximProshin

File: tests/torch/models_hub_test/test_timm.py [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+        name_set = set([n for n in _name.split("_") if not n.isnumeric()])
+        name_set = name_set.difference(ignore_set)
+        name_join = "_".join(name_set)
+        if name_join not in unique_models:
+            unique_models.add(name_join)
+            filtered_list.append(name)
+    return filtered_list
+
+
+def get_all_models() -> list:
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply tests/torch/models_hub_test/test_timm.py:44]: The comment has been reviewed and no further action will be taken. 
@MaximProshin

File: tests/torch/models_hub_test/test_torchvision_models.py [hunk 19]
```diff
@@ -0,0 +1,84 @@
+# Copyright (c) 2023 Intel Corporation
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply tests/torch/models_hub_test/test_torchvision_models.py:1]: The comment has been reviewed and no further action will be taken. 
@MaximProshin

File: tests/torch/models_hub_test/test_speech_transformer.py [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+import subprocess
+import sys
+import tempfile
+
+import torch
+
+from tests.torch.models_hub_test.common import BaseTestModel
+
+
+class TestSpeechTransformerModel(BaseTestModel):
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply tests/torch/models_hub_test/test_speech_transformer.py:25]: The comment has been reviewed and no further action will be taken. 
@MaximProshin

File: tests/torch/models_hub_test/test_timm.py [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+class TestTimmModel(BaseTestModel):
+    def load_model(self, model_name: str):
+        m = timm.create_model(model_name, pretrained=False)
+        cfg = timm.get_pretrained_cfg(model_name)
+        shape = [1] + list(cfg.input_size)
+        example = (torch.randn(shape),)
+        return m, example
+
+    @pytest.mark.parametrize("name", get_all_models())
+    def test_nncf_wrap(self, name):
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply tests/torch/models_hub_test/test_timm.py:58]: The comment has been reviewed and no further action will be taken. 
@MaximProshin

File: tests/torch/models_hub_test/test_timm.py [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+# limitations under the License.
+
+import pytest
+import timm
+import torch
+
+from tests.torch.models_hub_test.common import BaseTestModel
+
+
+def filter_timm(timm_list: list) -> list:
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply tests/torch/models_hub_test/test_timm.py:23]: The comment has been reviewed and no further action will be taken. 
@MaximProshin

**Pull request merged**